Sources/src/FromSTMTestBeamData_source.cc

```diff
@@ -0,0 +1,335 @@
+  {
+    using Name = fhicl::Name;
+    using Comment = fhicl::Comment;
+    fhicl::Sequence<std::string> inputFiles{Name("fileNames"),Comment("Input binary file")};
+    fhicl::Atom<unsigned int> maxEvents{Name("maxEvents"), Comment("Max number of events")};
+    fhicl::Atom<int> verbosityLevel{Name("verbosityLevel"), Comment("Verbosity level")};
+
+    // These are used by art and are required.
+    fhicl::Atom<std::string> module_label{Name("module_label"), Comment("Art module label"), ""};
+    fhicl::Atom<std::string> module_type{Name("module_type"), Comment("Art module type"), ""};
+  };
+  typedef fhicl::WrappedTable<Config> Parameters;
+
+
+  //================================================================
+  class STMTestBeamDataDetail : private boost::noncopyable {
+    std::string myModuleLabel_;
+    art::SourceHelper const& pm_;
+    unsigned runNumber_; // from file name
+    art::SubRunID lastSubRunID_;
+    std::set<art::SubRunID> seenSRIDs_;
+
+    unsigned currentFileNumber_;
+    std::string currentFileName_;
+    std::ifstream* currentFile_ = nullptr;
+
+    unsigned currentSubRunNumber_; // from file
+    unsigned currentEventNumber_;
+    unsigned maxEvents_;
+    int verbosityLevel_;
+    uint16_t channel_; // the channel (HPGe or LaBr) will be different for each file
+    int binaryFileVersion_;
+
+    int printAtEvent;
+
+    // A helper function used to manage the principals.
+    // This is boilerplate that does not change if you change the data products.
+    void managePrincipals ( int runNumber,
+        int subRunNumber,
+        int eventNumber,
+        art::RunPrincipal*&    outR,
+        art::SubRunPrincipal*& outSR,
+        art::EventPrincipal*&  outE);
+
+    public:
+    STMTestBeamDataDetail(const Parameters &conf,
+        art::ProductRegistryHelper &,
+        const art::SourceHelper &);
+
+    void readFile(std::string const& filename, art::FileBlock*& fb);
+
+    bool readNext(art::RunPrincipal* const& inR,
+        art::SubRunPrincipal* const& inSR,
+        art::RunPrincipal*& outR,
+        art::SubRunPrincipal*& outSR,
+        art::EventPrincipal*& outE);
+
+    void closeCurrentFile();
+  };
+
+  //----------------------------------------------------------------
+  STMTestBeamDataDetail::STMTestBeamDataDetail(const Parameters& conf,
+      art::ProductRegistryHelper& rh,
+      const art::SourceHelper& pm)
+    : myModuleLabel_("FromSTMTestBeamData")
+    , pm_(pm)
+    , runNumber_(0)
+    , currentFileNumber_(0)
+    , currentSubRunNumber_(-1U)
+    , currentEventNumber_(0)
+    , maxEvents_(conf().maxEvents())
+    , verbosityLevel_(conf().verbosityLevel())
+    , binaryFileVersion_(0)
+  {
+    rh.reconstitutes<mu2e::STMWaveformDigiCollection,art::InEvent>(myModuleLabel_, "HPGe");
+    rh.reconstitutes<mu2e::STMWaveformDigiCollection,art::InEvent>(myModuleLabel_, "LaBr");
+    rh.reconstitutes<mu2e::STMTestBeamEventInfo,art::InEvent>(myModuleLabel_);
+
+    currentSubRunNumber_ = 0;
```

Comment:
These lines, and a lot in the initialize list can go away if you initialize in the declaration.

Reply:
Done in [86c7106](https://github.com/Mu2e/Offline/pull/856/commits/86c7106ce0d9319e26f96d34d4fdbda49c37ab29)